src/annotator/anchoring/pdf.js

```diff
@@ -0,0 +1,430 @@
+ * @param {TextQuoteSelector} quoteSelector
+ * @return {Promise<Range>} Location of quote
+ */
+function findInPages(pageIndexes, quoteSelector, positionHint) {
+  if (pageIndexes.length === 0) {
+    return Promise.reject(new Error('Quote not found'));
+  }
+
+  const [pageIndex, ...rest] = pageIndexes;
+
+  const content = getPageTextContent(pageIndex);
+  const page = getPage(pageIndex);
+  const offset = getPageOffset(pageIndex);
+
+  const attempt = ([, content, offset]) => {
+    const root = { textContent: content };
+    const anchor = TextQuoteAnchor.fromSelector(root, quoteSelector);
+    if (positionHint) {
+      let hint = positionHint.start - offset;
+      hint = Math.max(0, hint);
+      hint = Math.min(hint, content.length);
+      return anchor.toPositionAnchor({ hint });
+    }
+    return anchor.toPositionAnchor();
+  };
+
+  const next = () => findInPages(rest, quoteSelector, positionHint);
+
+  const cacheAndFinish = anchor => {
+    if (positionHint) {
+      if (!quotePositionCache[quoteSelector.exact]) {
+        quotePositionCache[quoteSelector.exact] = {};
+      }
+      quotePositionCache[quoteSelector.exact][positionHint.start] = {
+        page,
+        anchor,
+      };
+    }
+    return anchorByPosition(page, anchor);
+  };
+
+  return Promise.all([page, content, offset])
+    .then(attempt)
+    .then(cacheAndFinish)
+    .catch(next);
+}
+
+/**
+ * Return a list of page indexes to search for a quote in priority order.
+ *
+ * When a position anchor is available, quote search can be optimized by
+ * searching pages nearest the expected position first.
+ *
+ * @param [TextPositionAnchor] position
+ * @return {number[]}
+ */
+function prioritizePages(position) {
+  const pageCount = PDFViewerApplication.pdfViewer.pagesCount;
+  const pageIndices = Array(pageCount)
+    .fill(0)
+    .map((_, i) => i);
+
+  if (!position) {
+    return Promise.resolve(pageIndices);
+  }
+
+  // Sort page indexes by offset from `pageIndex`.
+  function sortPages(pageIndex) {
+    const left = pageIndices.slice(0, pageIndex);
+    const right = pageIndices.slice(pageIndex);
+    const result = [];
+    while (left.length > 0 || right.length > 0) {
+      if (right.length) {
+        result.push(right.shift());
+      }
+      if (left.length) {
+        result.push(left.pop());
+      }
+    }
+    return result;
+  }
+
+  return findPage(position.start).then(({ index }) => sortPages(index));
+}
+
+/**
+ * Anchor a set of selectors to a DOM Range.
+ *
+ * @param {HTMLElement} root
+ * @param {Array} selectors - Selector objects to anchor
+ * @param {Object} options - Options to pass to selector anchoring
+ * @return {Promise<Range>}
+ */
+function anchor(root, selectors, options = {}) {
+  const position = selectors.find(s => s.type === 'TextPositionSelector');
+  const quote = selectors.find(s => s.type === 'TextQuoteSelector');
+
+  let result = Promise.reject('unable to anchor');
```

Comment:
Note for later refactoring: this is a bit of an unidiomatic use of Promise rejections.

Reply:
Indeed! I avoided changing the API in this PR in case some other part of the code relies on getting a `string` rather than an `Error` back.

Reply:
Agreed. I didn't want to change the public API in this PR though to avoid needing to make changes in other modules.